File: Packages/PassthroughCore/Sources/PassthroughCore/ControlServer.swift
import Foundation
import Network

/// Facts about the phone the Mac likes to display.
public struct DeviceStatus: Sendable, Equatable {
    public var deviceName: String
    public var radio: String?
    public var carrier: String?
    public var battery: Double?
    public var hosting: String
    /// Whether the network the phone sends Mac traffic out on routes IPv6; nil when unknown.
    public var ipv6: Bool?
    public init(deviceName: String, radio: String? = nil, carrier: String? = nil, battery: Double? = nil, hosting: String, ipv6: Bool? = nil) {
        self.deviceName = deviceName; self.radio = radio; self.carrier = carrier; self.battery = battery; self.hosting = hosting; self.ipv6 = ipv6
    }
}

/// A connected Mac as seen by the phone.
public struct ConnectedMac: Identifiable, Sendable, Equatable, Codable {
    public let id: String
    public let name: String
    public let since: Date
    public init(id: String, name: String, since: Date) { self.id = id; self.name = name; self.since = since }
}

/// Newline-delimited JSON control channel: pairing, heartbeat, live status.
public final class ControlServer: @unchecked Sendable {
    public let port: UInt16
    private let registry: PairingRegistry
    private let counter: ByteCounter
    private let statusProvider: @Sendable () -> DeviceStatus
    private let socksPort: UInt16
    fileprivate let queue = DispatchQueue(label: "dev.dpatel.passthrough.control")
    private var listeners: [NWListener] = []
    private var peers: [ObjectIdentifier: Peer] = [:]
    public private(set) var isRunning = false
    public var onClientsChanged: (@Sendable ([ConnectedMac]) -> Void)?
    /// A Mac was just linked for the wireless link, on the control queue.
    public var onLinked: (@Sendable () -> Void)?

    public init(port: UInt16 = PassthroughProtocol.defaultControlPort,
                socksPort: UInt16,
                registry: PairingRegistry,
                counter: ByteCounter,
                statusProvider: @escaping @Sendable () -> DeviceStatus) {
        self.port = port
        self.socksPort = socksPort
        self.registry = registry
        self.counter = counter
        self.statusProvider = statusProvider
    }

    public func start() throws {
        try queue.sync {
            guard !isRunning else { return }
            var lastError: Error?
            for host in [NWEndpoint.Host.ipv4(.loopback), .ipv6(.loopback)] {
                do { listeners.append(try makeListener(host: host)) } catch { lastError = error }
            }
            if listeners.isEmpty, let lastError { throw lastError }
            isRunning = true
            ptLog(.info, "Control channel listening on port \(port)")
        }
    }

    public func stop() {
        queue.sync {
            guard isRunning else { return }
            isRunning = false
            listeners.forEach { $0.cancel() }
            listeners = []
            peers.values.forEach { $0.cancel() }
            peers.removeAll()
            onClientsChanged?([])
        }
    }

    public var connectedMacs: [ConnectedMac] {
        queue.sync { authenticatedMacs() }
    }

    private func authenticatedMacs() -> [ConnectedMac] {
        peers.values.compactMap { $0.mac }.sorted { $0.since < $1.since }
    }

    private func makeListener(host: NWEndpoint.Host) throws -> NWListener {
        let params = NWParameters.tcp
        params.allowLocalEndpointReuse = true
        params.acceptLocalOnly = true
        params.requiredLocalEndpoint = .hostPort(host: host, port: NWEndpoint.Port(rawValue: port)!)
        let listener = try NWListener(using: params)
        listener.newConnectionHandler = { [weak self] connection in
            guard let self else { connection.cancel(); return }
            guard self.peers.count < 8 else { connection.cancel(); return }   // a handful of Macs, not a flood
            let peer = Peer(server: self, connection: connection)
            self.peers[ObjectIdentifier(peer)] = peer
            peer.start(on: self.queue)
        }
        listener.start(queue: queue)
        return listener
    }

    fileprivate func remove(_ peer: Peer) {
        peers[ObjectIdentifier(peer)] = nil
        onClientsChanged?(authenticatedMacs())
    }

    fileprivate func clientsChanged() {
        onClientsChanged?(authenticatedMacs())
    }

    // MARK: Message handling

    fileprivate func handle(_ message: ControlEnvelope, from peer: Peer) -> [ControlEnvelope] {
        switch message.t {
        case ControlEnvelope.hello:
            guard (message.protocolVersion ?? 0) == PassthroughProtocol.version else {
                var e = ControlEnvelope(t: ControlEnvelope.error); e.reason = PairingFailure.unsupportedVersion.rawValue
                return [e]
            }
            var paired = false
            if let id = message.clientID, let token = message.token, registry.verify(clientID: id, token: token) {
                paired = true
                registry.touch(clientID: id)
                peer.authenticate(id: id, name: message.name ?? "Mac")
            }
            var reply = welcome()
            reply.paired = paired
            return [reply]
        case ControlEnvelope.pair:
            guard let id = message.clientID, let code = message.code else {
                var e = ControlEnvelope(t: ControlEnvelope.error); e.reason = PairingFailure.badCode.rawValue
                return [e]
            }
            switch registry.pair(code: code, clientID: id, name: message.name ?? "Mac") {
            case .success(let token):
                peer.authenticate(id: id, name: message.name ?? "Mac")
                var reply = ControlEnvelope(t: ControlEnvelope.paired)
                reply.token = token
                reply.deviceName = statusProvider().deviceName
                reply.socksPort = Int(socksPort)
                return [reply, statusMessage()]
            case .failure(let failure):
                var e = ControlEnvelope(t: ControlEnvelope.error); e.reason = failure.rawValue
                return [e]
            }
        case ControlEnvelope.link:
            // Only a Mac that already proved its pairing token may link.
            guard let mac = peer.mac, let cert = message.certSHA256, cert.count == 64,
                  let key = message.linkKey.flatMap({ Data(base64Encoded: $0) }), key.count == 32 else { return [] }
            // No reply if the key could not be stored: the Mac stays unlinked
            // and offers the link again next time.
            guard registry.link(clientID: mac.id, certificateSHA256: cert, linkKey: key) else {
                ptLog(.error, "Could not link \(mac.name) for the wireless link")
                return []
            }
            var reply = ControlEnvelope(t: ControlEnvelope.linked)
            reply.phoneID = registry.phoneID
            ptLog(.info, "Linked \(mac.name) for the wireless link")
            onLinked?()
            return [reply]
        case ControlEnvelope.ping:
            return [ControlEnvelope(t: ControlEnvelope.pong)]
        default:
            return []
        }
    }

    private func welcome() -> ControlEnvelope {
        let status = statusProvider()
        var reply = ControlEnvelope(t: ControlEnvelope.welcome)
        reply.protocolVersion = PassthroughProtocol.version
        reply.deviceName = status.deviceName
        reply.socksPort = Int(socksPort)
        reply.radio = status.radio
        reply.carrier = status.carrier
        reply.battery = status.battery
        reply.hosting = status.hosting
        reply.ipv6 = status.ipv6
        return reply
    }

    fileprivate func statusMessage() -> ControlEnvelope {
        let status = statusProvider()
        let snap = counter.snapshot()
        var m = ControlEnvelope(t: ControlEnvelope.status)
        m.deviceName = status.deviceName
        m.radio = status.radio
        m.carrier = status.carrier
        m.battery = status.battery
        m.hosting = status.hosting
        m.ipv6 = status.ipv6
        m.activeConnections = snap.active
        m.rxBytes = snap.rx
        m.txBytes = snap.tx
        m.timestamp = Date().timeIntervalSince1970
        return m
    }
}

private final class Peer: @unchecked Sendable {
    private let server: ControlServer
    private let channel: ControlConnection
    private var timer: DispatchSourceTimer?
    private var cancelled = false
    private(set) var mac: ConnectedMac?

    init(server: ControlServer, connection: NWConnection) {
        self.server = server
        self.channel = ControlConnection(connection, queue: server.queue)
    }

    func start(on queue: DispatchQueue) {
        channel.onMessage = { [weak self] message in
            guard let self else { return }
            self.channel.send(self.server.handle(message, from: self))
        }
        channel.onClose = { [weak self] _ in self?.cancel() }
        channel.start(queue: queue)
        // Status broadcast once authenticated.
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + 1, repeating: 1)
        timer.setEventHandler { [weak self] in
            guard let self, self.mac != nil else { return }
            self.channel.send(self.server.statusMessage())
        }
        timer.resume()
        self.timer = timer
    }

    func authenticate(id: String, name: String) {
        if mac == nil {
            mac = ConnectedMac(id: id, name: name, since: Date())
            ptLog(.info, "\(name) connected")
            server.clientsChanged()
        }
    }

    func cancel() {
        guard !cancelled else { return }
        cancelled = true
        timer?.cancel()
        channel.cancel()
        if let mac { ptLog(.info, "\(mac.name) disconnected") }
        server.remove(self)
    }
}

File: android/app/src/main/java/dev/dpatel/passthrough/core/ControlServer.kt
package dev.dpatel.passthrough.core

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Newline-delimited JSON control channel: pairing, heartbeat, live status.
 * Mirrors PassthroughCore/ControlServer.swift message for message.
 */
class ControlServer(
    val port: Int = PassthroughProtocol.DEFAULT_CONTROL_PORT,
    private val socksPort: Int,
    private val registry: PairingRegistry,
    private val counter: ByteCounter,
    private val bindAddress: InetAddress = InetAddress.getLoopbackAddress(),
    private val statusProvider: () -> DeviceStatus,
) {
    @Volatile var onClientsChanged: ((List<ConnectedMac>) -> Unit)? = null
    private val peers = ConcurrentHashMap.newKeySet<Peer>()
    @Volatile private var serverSocket: ServerSocket? = null
    private var ticker: ScheduledExecutorService? = null
    @Volatile var isRunning = false
        private set

    val boundPort: Int get() = serverSocket?.localPort ?: port

    @Synchronized
    fun start() {
        if (isRunning) return
        val ss = ServerSocket()
        ss.reuseAddress = true
        ss.bind(InetSocketAddress(bindAddress, port), 16)
        serverSocket = ss
        isRunning = true
        Thread({ acceptLoop(ss) }, "control-accept").apply { isDaemon = true }.start()
        // Status broadcast once a second to every authenticated Mac.
        ticker = Executors.newSingleThreadScheduledExecutor { r -> Thread(r, "control-status").apply { isDaemon = true } }.also {
            it.scheduleWithFixedDelay({
                val authenticated = peers.filter { p -> p.mac != null }
                if (authenticated.isNotEmpty()) {
                    val status = statusMessage()
                    authenticated.forEach { p -> p.send(listOf(status)) }
                }
            }, 1, 1, TimeUnit.SECONDS)
        }
        ptLog(PtLog.Level.INFO, "Control channel listening on port ${ss.localPort}")
    }

    @Synchronized
    fun stop() {
        if (!isRunning) return
        isRunning = false
        ticker?.shutdownNow(); ticker = null
        runCatching { serverSocket?.close() }
        serverSocket = null
        peers.toList().forEach { it.cancel() }
        peers.clear()
        onClientsChanged?.invoke(emptyList())
    }

    val connectedMacs: List<ConnectedMac> get() = peers.mapNotNull { it.mac }.sortedBy { it.since }

    private fun acceptLoop(ss: ServerSocket) {
        while (isRunning) {
            val socket = try { ss.accept() } catch (_: IOException) { break }
            if (peers.size >= 8) { runCatching { socket.close() }; continue }   // a handful of Macs, not a flood
            val peer = Peer(socket)
            peers.add(peer)
            Thread({ peer.run() }, "control-peer").apply { isDaemon = true }.start()
        }
    }

    // Message handling

    private fun handle(m: ControlEnvelope, peer: Peer): List<ControlEnvelope> = when (m.t) {
        ControlEnvelope.HELLO -> {
            if ((m.protocolVersion ?: 0) != PassthroughProtocol.VERSION) {
                listOf(ControlEnvelope(ControlEnvelope.ERROR, reason = PairingFailure.UNSUPPORTED_VERSION.wire))
            } else {
                val id = m.clientID
                val token = m.token
                var paired = false
                if (id != null && token != null && registry.verify(id, token)) {
                    paired = true
                    registry.touch(id)
                    peer.authenticate(id, m.name ?: "Mac")
                }
                listOf(welcome().apply { this.paired = paired })
            }
        }
        ControlEnvelope.PAIR -> {
            val id = m.clientID
            val code = m.code
            if (id == null || code == null) {
                listOf(ControlEnvelope(ControlEnvelope.ERROR, reason = PairingFailure.BAD_CODE.wire))
            } else when (val r = registry.pair(code, id, m.name ?: "Mac")) {
                is PairingRegistry.PairResult.Success -> {
                    peer.authenticate(id, m.name ?: "Mac")
                    listOf(
                        ControlEnvelope(ControlEnvelope.PAIRED, token = r.token, deviceName = statusProvider().deviceName, socksPort = socksPort),
                        statusMessage(),
                    )
                }
                is PairingRegistry.PairResult.Failure -> listOf(ControlEnvelope(ControlEnvelope.ERROR, reason = r.reason.wire))
            }
        }
        ControlEnvelope.PING -> listOf(ControlEnvelope(ControlEnvelope.PONG))
        else -> emptyList()
    }

    private fun welcome(): ControlEnvelope {
        val s = statusProvider()
        return ControlEnvelope(
            ControlEnvelope.WELCOME, protocolVersion = PassthroughProtocol.VERSION, deviceName = s.deviceName, socksPort = socksPort,
            radio = s.radio, carrier = s.carrier, battery = s.battery, hosting = s.hosting,
        )
    }

    private fun statusMessage(): ControlEnvelope {
        val s = statusProvider()
        val snap = counter.snapshot()
        return ControlEnvelope(
            ControlEnvelope.STATUS, deviceName = s.deviceName, radio = s.radio, carrier = s.carrier, battery = s.battery,
            hosting = s.hosting, activeConnections = snap.active, rxBytes = snap.rx, txBytes = snap.tx,
            timestamp = System.currentTimeMillis() / 1000.0,
        )
    }

    private inner class Peer(private val socket: Socket) {
        private val cancelled = AtomicBoolean(false)
        private lateinit var out: OutputStream
        @Volatile var mac: ConnectedMac? = null
            private set

        fun run() {
            try {
                socket.tcpNoDelay = true
                out = socket.getOutputStream()
                val input = socket.getInputStream()
                while (!cancelled.get()) {
                    val line = readLine(input) ?: break
                    if (line.isBlank()) continue
                    val message = try { ControlEnvelope.decode(line) } catch (_: Exception) {
                        ptLog(PtLog.Level.WARNING, "control: undecodable message"); continue
                    }
                    send(handle(message, this))
                }
            } catch (_: IOException) {
            } finally {
                cancel()
            }
        }

        /** Reads one newline-terminated line, capped so a peer can't balloon memory. */
        private fun readLine(input: InputStream): String? {
            val buf = ByteArrayOutputStream()
            while (true) {
                val b = input.read()
                if (b < 0) return null
                if (b == '\n'.code) return buf.toString(Charsets.UTF_8.name())
                buf.write(b)
                if (buf.size() > 256 * 1024) throw IOException("control line too long")
            }
        }

        fun authenticate(id: String, name: String) {
            if (mac == null) {
                mac = ConnectedMac(id, name, System.currentTimeMillis())
                ptLog(PtLog.Level.INFO, "$name connected over USB")
                onClientsChanged?.invoke(connectedMacs)
            }
        }

        fun send(messages: List<ControlEnvelope>) {
            if (messages.isEmpty() || cancelled.get() || !this::out.isInitialized) return
            val payload = ByteArrayOutputStream()
            messages.forEach { payload.write(it.encodeLine()) }
            try {
                synchronized(this) { out.write(payload.toByteArray()); out.flush() }
            } catch (_: IOException) {
                cancel()
            }
        }

        fun cancel() {
            if (!cancelled.compareAndSet(false, true)) return
            runCatching { socket.close() }
            peers.remove(this)
            mac?.let { ptLog(PtLog.Level.INFO, "${it.name} disconnected") }
            onClientsChanged?.invoke(connectedMacs)
        }
    }
}
